### PR TITLE
Add more functionality to tabs in honor of styleguide week

### PR DESCRIPTION
**Pull request by mmlumba**
References #1000 
screen shot 2017-03-09 at 4 26 49 pm

### CODE REVIEW

**Comment by shanear** [comment]
Is this still a WIP or should I review?

**Comment by mmlumba** [comment]
Its done now...please review

File: client/app/components/TabWindow.jsx [hunk 1]
```diff
@@ -26,29 +27,62 @@ export default class TabWindow extends React.Component {
...
+  }
+
   render() {
     let {
       tabs,
       pages,
       fullPage
     } = this.props;
 
+    let newTabs = [];
```
Comment (shanear) [inline client/app/components/TabWindow.jsx]: there is some weird stuff happening with variable scope and the `map` function here. Map will run over the array of tabs and return a new array with the property of the returned value. You are not using `map` that way, and it's sort of messy.

for instance:
```
let numbers = [1, 2, 3]

let bigNumbers = numbers.map((number) => {
  return number * 100
});

// bigNumbers will equal [100, 200, 300]
```

Does that make sense?

File: client/app/components/TabWindow.jsx [hunk 2]
```diff
@@ -26,29 +27,62 @@ export default class TabWindow extends React.Component {
...
+    });
+
     return <div>
         <div className={
           `cf-tab-navigation${fullPage ? " cf-tab-navigation-full-screen" : ""}`
         }>
-          {tabs.map((tab, i) =>
-            <div
-              className={`cf-tab${i === this.state.currentPage ? " cf-active" : ""}`}
+          {newTabs.map((tab, i) =>
```
Comment (shanear) [inline client/app/components/TabWindow.jsx]: again, I think you are intending to use `forEach` or `each` here, not `map`.

File: client/app/containers/StyleGuide/StyleGuideTabs.jsx [hunk 3]
```diff
@@ -27,29 +26,36 @@ export default class StyleGuideTabs extends React.Component {
...
-    tabHeadersWithIcons = [
-      " Active Tab",
-      "Enabled Tab",
-      "Disabled Tab"
+    tabHeadersWithIcons = [{
+      disable: false,
+      icon: successSymbolHtml(),
+      label: "Active Tab"
+    }, {
+      disable: false,
```
Comment (shanear) [inline client/app/containers/StyleGuide/StyleGuideTabs.jsx]: can we make it so they aren't disabled by default? If that is the case, we shouldn't have to pass `disable: false` here

File: client/app/components/TabWindow.jsx [hunk 4]
```diff
@@ -26,29 +27,62 @@ export default class TabWindow extends React.Component {
...
     let {
       tabs,
       pages,
       fullPage
     } = this.props;
 
+    let newTabs = [];
+
+
+    tabs.map((tab) => {
```
Comment (shanear) [inline client/app/components/TabWindow.jsx]: Now that I've looked over everything. It seems like this step might be unnecessary. Can we just loop through the `tabs` array in the JSX part and call `this.getTabHeaderWithSVG(tab)` there?

File: client/app/components/TabWindow.jsx [hunk 5]
```diff
@@ -26,29 +27,62 @@ export default class TabWindow extends React.Component {
...
+              onClick={tab.disable ? this.onDisabledTabClick(i) : this.onTabClick(i)}
+              disabled={Boolean(tab.disable)}>
               <span>
-                <span>
-                  {tab}
-                </span>
+                {tab.element}
               </span>
-            </div>
+            </button>
```
Comment (shanear) [inline client/app/components/TabWindow.jsx]: yay accessibility :)

File: Gemfile.lock [hunk 6]
```diff
@@ -106,13 +106,13 @@ GEM
       nokogiri
     arel (6.0.4)
     ast (2.3.0)
-    aws-sdk (2.8.3)
-      aws-sdk-resources (= 2.8.3)
-    aws-sdk-core (2.8.3)
+    aws-sdk (2.8.4)
```
Comment (mmlumba) [inline Gemfile.lock]: Should we explicitly lock versions of the gems that will be affected whenever we make changes for commons in the Gemfile?

File: client/app/components/TabWindow.jsx [hunk 7]
```diff
@@ -26,29 +27,62 @@ export default class TabWindow extends React.Component {
...
       tabs,
       pages,
       fullPage
     } = this.props;
 
+    let newTabs = [];
+
+
+    tabs.map((tab) => {
+      newTabs.push({ disable: tab.disable,
```
Comment (mmlumba) [inline client/app/components/TabWindow.jsx]: `disable` is a prop that sets the HTML attribute `disabled`; if `disable` is true add the `disabled` prop like so: `<button className="cf-tab" disabled></button>`

File: client/app/containers/StyleGuide/StyleGuideTabs.jsx (hunk 3 above)
Comment (mmlumba) [inline client/app/containers/StyleGuide/StyleGuideTabs.jsx]: The `disable: false` makes it so that the `disabled` attribute isn't passed through the HTML tags of the tabs

File: client/app/components/TabWindow.jsx [hunk 8]
```diff
@@ -26,29 +27,62 @@ export default class TabWindow extends React.Component {
     }
   }
 
+  onDisabledTabClick = (tab) => () => {
```
Comment (mmlumba) [inline client/app/components/TabWindow.jsx]: This adds the `disabled` attribute to the cf-tab element; US web standards designs the `disabled` buttons through a `disabled` attribute so that's how I styled the tab.

File: Gemfile.lock (hunk 6 above)
Comment (shanear) [inline Gemfile.lock]: no, cause we do want these updates, they should just be in their own PR so we aren't making multiple changes at once

File: client/app/components/TabWindow.jsx (hunk 8 above)
Comment (shanear) [inline client/app/components/TabWindow.jsx]: This is a click handler, I might be wrong but you shouldn't need this to make a tab disabled. Maybe lets pair on this.

File: client/app/containers/StyleGuide/StyleGuideTabs.jsx (hunk 3 above)
Comment (shanear) [inline client/app/containers/StyleGuide/StyleGuideTabs.jsx]: right, but it should default to false, so we don't have to pass it.

**Comment by mmlumba** [comment]
Posted question on the engineering channel. My main concern about that approach is that while something like that would work for a simple HTML page, this may not work for something more complex (such as putting in another React component for each page or using a function to change the page content).

**Comment by mmlumba** [comment]
@shanear @mdbenjam Can you take a final look? Also adding Mark since this change also affects Dispatch functionality.

File: client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx [hunk 9]
```diff
@@ -306,11 +304,10 @@ export default class EstablishClaimDecision extends React.Component {
             <p>Use the tabs to review the decision documents below and
             select the decision that best fits the VACOLS Decision Criteria.</p>
             <TabWindow
-              tabs={tabHeaders}
-              pages={pdfViews}
+              tabs={tabs}
               onChange={this.onTabSelected}/>
           </div>}
-          {!this.hasMultipleDecisions() && pdfViews[0]}
+          {!this.hasMultipleDecisions() && tabs[0].page}
```
Comment (shanear) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: shouldn't this check be inside of the tab `page`?

Comment (shanear) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: shouldn't this check be inside of the tab page?

File: client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx [hunk 10]
```diff
@@ -220,48 +220,46 @@ export default class EstablishClaimDecision extends React.Component {
...
-                className="cf-doc-embed cf-iframe-with-loading"
-                title="Form8 PDF"
-                src={`${pdfjsLink}&decision_number=${index}`}>
-              </iframe>
-            </LoadingContainer>
-          </div>
-        </div>);
+      tab.label = `Decision ${(index + 1)} (${formatDate(decision.received_at)})`;
+
+      tab.page = <div>
```
Comment (shanear) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: We should consider refactoring this into it's own component. But that can come for later.

File: client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx (hunk 9 above)
Comment (mmlumba) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: This check renders one page only if there aren't multiple decisions.

Comment (shanear) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: ohh, this is an if/else. Is there a better way to add this to the conditional in line 302?

Comment (mmlumba) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: The conditional in 302-309 shows the message for selecting the decision document as well as the tab component if there are multiple decisions.

Comment (shanear) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: turns out this is actually a thing people do in React 😝 

my bad

File: client/app/components/RenderFunctions.jsx [hunk 11]
```diff
@@ -114,5 +114,47 @@ export let missingSymbolHtml = function() {
   );
 };
 
+export let checkSymbolHtml = function() {
```
Comment (mdbenjam) [inline client/app/components/RenderFunctions.jsx]: What do you think about using Font Awesome instead of custom SVGs. I'm cool with an SVG if it's offering something new, but we've added Font Awesome, so it should be easy to use something like: http://fontawesome.io/icon/check/

File: client/app/components/RenderFunctions.jsx [hunk 12]
```diff
@@ -114,5 +114,47 @@ export let missingSymbolHtml = function() {
...
+        L4.02575165,12.3545702 L5.42970942,13.7297517 C5.62229274,13.9185289 5.85630682,
+        14.0128554 6.13162495,14.0128554 C6.40681639,14.0128554 6.64083046,13.9184047
+        6.83354049,13.7297517 L8.23749825,12.3545702 L15.7109982,5.03359071 C15.9035815,
+        4.84481355 15.9999999,4.61557527 15.9999999,4.34587587 C16.0001266,4.07630059
+        15.9035815,3.84706231 15.7109982,3.65828514 L14.3071671,2.28310369 Z" id="Shape"></path>
+  </svg>
+  );
+};
+
+export let crossSymbolHtml = function() {
```
Comment (mdbenjam) [inline client/app/components/RenderFunctions.jsx]: Is this an x shape? Maybe use http://fontawesome.io/icon/times/

File: client/app/components/TabWindow.jsx [hunk 13]
```diff
@@ -38,28 +53,26 @@ export default class TabWindow extends React.Component {
...
         </div>
       </div>;
   }
 }
 
 TabWindow.propTypes = {
   onChange: PropTypes.func,
-  pages: PropTypes.arrayOf(PropTypes.node).isRequired,
-  tabs: PropTypes.arrayOf(PropTypes.string).isRequired
+  tabs: PropTypes.arrayOf(PropTypes.object).isRequired
```
Comment (mdbenjam) [inline client/app/components/TabWindow.jsx]: I recently learned about PropTypes.shape. I think we should start using that to be more explicit in our PropTypes, could you update this to show what type of object we should be passing?

Example:
```
React.PropTypes.shape({
    color: React.PropTypes.string,
    fontSize: React.PropTypes.number
})
```

File: client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx [hunk 14]
```diff
@@ -220,48 +220,46 @@ export default class EstablishClaimDecision extends React.Component {
...
 
-    let tabHeaders = decisions.map((decision, index) =>
-      `Decision ${(index + 1)} (${formatDate(decision.received_at)})`);
+    let tabs = decisions.map((decision, index) => {
+      let tab = {};
 
-    let pdfViews = decisions.map((decision, index) =>
+      tab.disable = false;
 
-      /* This link is here for 508 compliance, and shouldn't be visible to sighted
```
Comment (mdbenjam) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: Can we keep this comment in the new refactor?

File: client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx (hunk 9 above)
Comment (mdbenjam) [inline client/app/containers/EstablishClaimPage/EstablishClaimDecision.jsx]: Unfortunately you can't do an if else in jsx. We could factor it out above and use an if else for more clarity. Up to you.

**Comment by mdbenjam** [comment]
Looking good! Just a couple of comments. Also, it looks to me like the icons are not vertically centered, but are vertically aligned with the text. Which doesn't seem to be the case in the reference issue #1000. Is this an intentional shift?

**Comment by mmlumba** [comment]
@mdbenjam Gina specified that the icons should be vertically aligned with the text

File: client/app/components/TabWindow.jsx [hunk 15]
```diff
@@ -38,28 +53,31 @@ export default class TabWindow extends React.Component {
...
         </div>
       </div>;
   }
 }
 
 TabWindow.propTypes = {
   onChange: PropTypes.func,
-  pages: PropTypes.arrayOf(PropTypes.node).isRequired,
-  tabs: PropTypes.arrayOf(PropTypes.string).isRequired
+  tabs: PropTypes.arrayOf(PropTypes.shape({
```
Comment (mdbenjam) [inline client/app/components/TabWindow.jsx]: This makes me so happy :D. Thanks for adding it!!
